### ISSUE TITLE
Level order traversal of BST using queue

### ISSUE THREAD

**Issue by sonasingh46** [sonasingh46/data-structures#3]

**Comment by harshit23897** [comment]
Since there is no inbuilt queue in C, I am planning to add code for queue first. Can you say how to proceed. Thanks.

**Comment by sonasingh46** [comment]
I have added library for queue in the project. You can check it out.
You can use the same. @harshit23897

**Comment by harshit23897** [comment]
The implemented queue is not very generic and limited to integer data type only. This poses restriction to create queue of let's say nodes in BST.

**Comment by sonasingh46** [comment]
Correct point. But to begin we can use that as we will be doing the level order traversal of BST which will have integer data type only.
If you can make a generic queue library, feel free to make it my making a new folder e.g. `GenericQueue` and proceed. Your call. @harshit23897 
By the way thanks for showing interest. @harshit23897

**Comment by harshit23897** [comment]
I believe that is not correct since we need to store each node in queue for level order traversal. Correct me if I am wrong.

**Comment by sonasingh46** [comment]
Yes you need to store each node in queue. And the node will be of integer type only for the tree.
Is it a problem?

**Comment by harshit23897** [comment]
I think we need to store node, since in each iteration we'll be getting a node from queue and pushing all its children which are not yet visited by level order traversal into the queue. Hence, only storing the value of node will not suffice.

**Comment by sonasingh46** [comment]
Yes you are absolutely correct. We need to make the queue generic.. I somehow misunderstood your thoughts.
You can proceed by making the queue code generic or even write your own.
@harshit23897

**Comment by sonasingh46** [comment]
@harshit23897 
I have made issue to make each type of data structure generic.
You can pick any one and start implementing.

**Comment by sonasingh46** [comment]
@harshit23897 
https://github.com/sonasingh46/data-structures/pull/20

**Comment by harshit23897** [comment]
Sure, I will take this.

**Comment by harshit23897** [comment]
I looked into this matter but level order traversal cannot be done with current queue implementation. Since out current queue DS currently only has a variable of void type and a pointer to point on next queue element. Hence, we cannot a store a BST node in this queue (which has a variable of void type and two pointers for left as well as right child). This poses restriction for implementation of level order traversal using Queue.

**Comment by sonasingh46** [comment]
@harshit23897 
I was expecting this.
The main challenge here is to develop a generic data structure (like STL in C++, and generics in java)
I am trying to do that.(Not sure.... how much feasible)
Thanks for pointing it out.

**Comment by harshit23897** [comment]
I don't know much about implementation of generic data structures, but I believe it would be fairly complicated.
